### ISSUE TITLE
Allow installing standalone binaries from Toolpacks

### ISSUE THREAD

**Issue by 00sapo** [ivan-hc/AM#1079]
# Is your feature request related to a problem? Please describe.
Just expanding the number of packages provided

# Describe the solution you'd like
I see two options:
1. including the whole [Toolpacks](https://github.com/Azathothas/Toolpacks) list in the package list of AM
2. provide an option to act as package manager for [Toolpacks](https://github.com/Azathothas/Toolpacks) (this option is better suited if users don't trust Toolpacks)

**Comment by Samueru-sama** [comment]
I was talking with Ivan about this yesterday lol

**Comment by Samueru-sama** [comment]
> > I was talking with Ivan about this yesterday lol
> 
> I guess GitHub is pushing that repository in the recommendations of our home :P

We are actually in contact with @Azathothas 

He also has this [repo](https://github.com/pkgforge/pkgcache) for appimages and other portable formats. And there are two package managers that use it already [dbin](https://github.com/xplshn/dbin) and [soar](https://github.com/pkgforge/soar)

**Comment by ivan-hc** [comment]
Maybe I can add a `--toolpacks` flag to `-i`, without adding new options.

If enabled, it will check if the program exists, and if so, it will create an installation script on-the-fly to install the program and to made it updatable.

@Samueru-sama please provide me a direct link to a toolpack and a direct URL to an info page that contains the version of the program.

**Comment by Azathothas** [comment]
Hi @ivan-hc 

If you use `json/toml/yaml`, you can have all the information you are looking for, check:
https://github.com/pkgforge/pkgcache/blob/main/Docs/METADATA.md

However if you don't want to parse json, you can still download the binaries but you won't have access to any metadata like `version`, `description` or `homepage`
You also won't be able to view `build_logs` & `build_script` which might make the users question trust & security.
Regardless, here's how you can get direct downloads:
```bash
!#Static Binaries
wget "https://bin.pkgforge.dev/$(uname -m)/${BIN}"

#For example to download fastfetch
wget "https://bin.pkgforge.dev/$(uname -m)/fastfetch"

#It is not possible to get version without parsing metadata
```
```bash
!#AppImages,AppBundles,FlatImages etc
wget "https://pkg.pkgforge.dev/$(uname -m)/${PKG}.${FORMAT}"

#Examples:
wget "https://pkg.pkgforge.dev/$(uname -m)/steam.AppImage"
wget "https://pkg.pkgforge.dev/$(uname -m)/steam.FlatImage"

!#It is possible to get versions here directly, just add .version at end
curl -qfsSL "https://pkg.pkgforge.dev/$(uname -m)/steam.AppImage.version"
curl -qfsSL "https://pkg.pkgforge.dev/$(uname -m)/steam.FlatImage.version"
```

A very long (will take long time to load) Readme is here: 
- https://github.com/Azathothas/Toolpacks/blob/main/x86_64_Linux/DETAILED.md
- https://github.com/Azathothas/Toolpacks/blob/main/aarch64_arm64_Linux/DETAILED.md

For pkgs, it is here:
- https://github.com/pkgforge/pkgcache/blob/main/aarch64-Linux/DETAILED.min.md
- https://github.com/pkgforge/pkgcache/blob/main/x86_64-Linux/DETAILED.min.md

**Comment by ivan-hc** [comment]
Hi @Azathothas , I just asked:
- the URL to one app
- the URL to a file that contains the version

**Comment by ivan-hc** [comment]
> Hi @Azathothas , I just asked:
> 
>     * the URL to one app
> 
>     * the URL to a file that contains the version

I'll do the test with
```
wget https://pkg.pkgforge.dev/x86_64/steam.AppImage
curl -Ls https://pkg.pkgforge.dev/x86_64/steam.AppImage.version
```

**Comment by ivan-hc** [comment]
does `--appimage-extract` fork on flatimages?

**Comment by Samueru-sama** [comment]
> does `--appimage-extract` work on flatimages?

I don't think so and also it doesn't work on appbundles either.

Only NixAppImages have compatible flags I think.

**Comment by ivan-hc** [comment]
that said, I'll allow only appimages

**Comment by ivan-hc** [comment]
to undestand my issue with lowercased URLs, this is what happe if I remove the
```
arg=$(echo "$arg" | tr '[:upper:]' '[:lower:]' )
```
from that function, using this instead
```
_toolbox() {
	mkdir -p "$CACHEDIR/toolboxes"
	curl -Ls "$AMREPO"/templates/AM-SAMPLE-AppImage > "$CACHEDIR/toolboxes/$arg.sh" || exit 1
	sed -i "s#APP=SAMPLE#APP=$arg#g; \
	s#FUNCTION#curl -Ls https://pkg.pkgforge.dev/$ARCH/$arg.version#g; \
	s#wget \"\$version\"#wget https://pkg.pkgforge.dev/$ARCH/$arg#g" "$CACHEDIR/toolboxes/$arg.sh"
	chmod a+x "$CACHEDIR/toolboxes/$arg.sh"
	arg="$CACHEDIR/toolboxes/$arg.sh"
}
```

https://github.com/user-attachments/assets/f35f5a8f-8658-4a77-99ff-544ce18b9031

**Comment by Samueru-sama** [comment]
>will made, for example steam.AppImage lowercased, as steam.appimage, and this causes wget to have errors

I think this needs to be fixed upstream, because yeah when using `dbin` and `soar` the user also has to be careful with capitalization. 

However if you will only allow appimages, you can instead inject `https://pkg.pkgforge.dev/$ARCH/$arg.AppImage` having that hardcoded at the end.

**Comment by ivan-hc** [comment]
> However if you will only allow appimages, you can instead inject `https://pkg.pkgforge.dev/$ARCH/$arg.AppImage` having that hardcoded at the end.

this is a good idea

**Comment by ivan-hc** [comment]
anyone who wants to test and improve this function
```
am --devmode-enable
am -s
```

the syntax is this
```
am -i --user --toolpack {PROGRAM}
```
only AppImages are supported.

Have fun!

**Comment by ivan-hc** [comment]
https://github.com/user-attachments/assets/adeab97d-f201-47d3-8c44-18ccd633b758

**Comment by ivan-hc** [comment]
![Istantanea_2024-11-05_19-59-04 png](https://github.com/user-attachments/assets/a1bd81e0-bb0a-4c60-97de-1be66e2963ae)

**Comment by ivan-hc** [comment]
![Istantanea_2024-11-05_20-00-24 png](https://github.com/user-attachments/assets/c72fee69-242f-4df7-9062-2e1e076de495)

**Comment by ivan-hc** [comment]
Maybe, also `-a` should be redirected on that database

**Comment by ivan-hc** [comment]
@Azathothas how to get only the description and the URL reference of one app from the info?

I'm working on the info creation for `-a`
```
	elif echo "$package_name" | grep -q ".toolpack"; then
		toolpack_appimage_name=$(echo "$package_name" | sed 's/\.toolpack//g')
		toolpack_markdown_page="https://raw.githubusercontent.com/pkgforge/pkgcache/refs/heads/main/$ARCH-Linux/DETAILED.min.md"
		if curl -Ls "$toolpack_markdown_page" | grep -q "$toolpack_appimage_name"; then
			toolpack_description=
```

**Comment by ivan-hc** [comment]
I have solved with the description
```
	elif echo "$package_name" | grep -q ".toolpack$"; then
		toolpack_appimage_name=$(echo "$package_name" | sed 's/\.toolpack//g')
		toolpack_markdown_page="https://raw.githubusercontent.com/pkgforge/pkgcache/refs/heads/main/$ARCH-Linux/DETAILED.min.md"
		if curl -Ls "$toolpack_markdown_page" | grep -q "$toolpack_appimage_name"; then
			toolpack_description=$(curl -Ls "https://pkg.pkgforge.dev/$ARCH/86box.AppImage.repology.json" \
			| grep "\"summary\":" | head -1 | sed 's/"summary": //g; s/  //g; s/^"//g; s/",$/./g')
			toolpack_site=
	else
```
now I need a source

**Comment by ivan-hc** [comment]
This is the first draft of the function changed to show info about the single app installed
```
_about_show_markdown() {
	if curl -o /dev/null -sIf "$markdown_url"; then
		markdown_page=$(curl -Ls "$markdown_url" \
		| sed '1,${ /^\s*#/ d; /^\s*!/ d; /\[Applications]/d; /\ --- /d; /\ | - | /d; /\!\[/d; }' \
		| sed '$!N;s/^\s*\n\s*$//;P;D')
		echo "$markdown_page" | fold -sw 72 | sed 's/^ //g; s/^/ /g'
	elif echo "$package_name" | grep -q ".toolpack"; then
		toolpack_appimage_name=$(echo "$arg" | sed 's/\.toolpack//g')
		if curl -o /dev/null -sIf "https://pkg.pkgforge.dev/$ARCH/$toolpack_appimage_name.AppImage.repology.json"; then
			toolpack_description=""
			toolpack_site=""
			echo ""
			echo "$toolpack_description" | fold -sw 72 | sed 's/^ //g; s/^/ /g'
			echo ""
			echo "SOURCE: $toolpack_site" | fold -sw 72 | sed 's/^ //g; s/^/ /g'
		fi
	else
		printf ' "%s" IS NOT A VALID ARGUMENT\n' "$package_name"
	fi
}
```
...however, I giva up for the description and source

![Istantanea_2024-11-06_00-36-23 png](https://github.com/user-attachments/assets/395c2e47-37d7-49aa-bd99-0880c678df71)

**Comment by ivan-hc** [comment]
https://github.com/ivan-hc/AM/commit/a11505734930c3e38b6cdbed507201e55ea199b2

any contribution is wellcome

**Comment by ivan-hc** [comment]
For AppImages it is much more feasible to create on-the-fly install scripts... but not for other types of Toolpacks, since they may or may not require a launcher or icons. Automating the creation of these scripts like this is out of my reach for now (see option `-e`)... but I might give it a try anyway (if only the data provided in the project's JSON files were more easily... "decipherable"... and without `jq`).

**Comment by ivan-hc** [comment]
> For AppImages it is much more feasible to create on-the-fly install scripts... but not for other types of Toolpacks, since they may or may not require a launcher or icons. Automating the creation of these scripts like this is out of my reach for now (see option `-e`)... but I might give it a try anyway (if only the data provided in the project's JSON files were more easily... "decipherable"... and without `jq`).

> > will made, for example steam.AppImage lowercased, as steam.appimage, and this causes wget to have errors
> 
> I think this needs to be fixed upstream, because yeah when using `dbin` and `soar` the user also has to be careful with capitalization.
> 
> However if you will only allow appimages, you can instead inject `https://pkg.pkgforge.dev/$ARCH/$arg.AppImage` having that hardcoded at the end.

@Samueru-sama this kind of approach (currently implemented in "dev") favors AppImage but prevents other formats. @Azathothas I find the indexing of application links a bit "confusing" for management via AM. So either we leave this function only for AppImage and increase the functions for other formats, or we unify the function so that it can handle both.

**Comment by ivan-hc** [comment]
???

![Istantanea_2024-11-06_02-58-21 png](https://github.com/user-attachments/assets/88d3255d-04e9-41e9-be9e-24ebabea4bb1)

**Comment by ivan-hc** [comment]
just noticed that this page contain all the info I need also for `-a` https://raw.githubusercontent.com/pkgforge/pkgcache/refs/heads/main/x86_64-Linux/README.md and maybe I can extend it to `-l`

**Comment by Azathothas** [comment]
Instead of trying to parse markdown and unreliable files, please consider using the json/toml/yaml.
You can use your favorite sed and grep if using jq isn't practical.
It is early here, however I will try showing you how to use sed/grep to do what jq does.

All of the markdown files you are using is autogenerated and may change at any moment since it's out of our control.

If you use the metadata, it solves all the problems you are encountering.

The vlc version is parsed from their git tags, so if the git tags itself has incorrect info, it will show up like that. 
However I should add a check there to ignore it if alphanumeric chars aren't found.
It's likely that vlc isn't the only app that has incorrect version, thanks for catching it.

**Comment by ivan-hc** [comment]
In the meantime I've solved for the info of the AppImages
```
_about_show_markdown() {
	if curl -o /dev/null -sIf "$markdown_url"; then
		markdown_page=$(curl -Ls "$markdown_url" \
		| sed '1,${ /^\s*#/ d; /^\s*!/ d; /\[Applications]/d; /\ --- /d; /\ | - | /d; /\!\[/d; }' \
		| sed '$!N;s/^\s*\n\s*$//;P;D')
		echo "$markdown_page" | fold -sw 72 | sed 's/^ //g; s/^/ /g'
	elif echo "$package_name" | grep -q ".toolpack"; then
		toolpack_appimage_name=$(echo "$arg" | sed 's/\.toolpack//g')
		if curl -o /dev/null -sIf "https://pkg.pkgforge.dev/$ARCH/$toolpack_appimage_name.AppImage.repology.json"; then
			toolpack_readme="https://raw.githubusercontent.com/pkgforge/pkgcache/refs/heads/main/$ARCH-Linux/README.md"
			toolpack_description=$(curl -Ls "$toolpack_readme" | grep -i "$toolpack_appimage_name" | sed 's/ | /\n/g' | head -5 | tail -1)
			toolpack_site=$(curl -Ls "$toolpack_readme" | grep -i "$toolpack_appimage_name" | sed 's/ | /\n/g' | tr '[()]' '\n' | grep "^http" | head -2 | tail -1)
			printf "\n%b\n\nSOURCE: %b\n" "$toolpack_description" "$toolpack_site" | fold -sw 72 | sed 's/^ //g; s/^/ /g'
		fi
	else
		printf ' "%s" IS NOT A VALID ARGUMENT\n' "$package_name"
	fi
}
```
![Istantanea_2024-11-06_03-40-16 png](https://github.com/user-attachments/assets/cc374f8d-9869-4128-b485-3af2661755bb)

**Comment by ivan-hc** [comment]
since the appImage is solved, now we need to finalize for all the other programs... and honestly, I have no idea how to start. Since you are more familiar with AppBundle and other formats... you could start a study on the changes I made in the "dev" branch and start adding the missing code.

https://github.com/ivan-hc/AM/pull/1082/files

**Comment by Azathothas** [comment]
> that said, I'll allow only appimages

I believe @00sapo wanted the static binaries and he wasn't even aware that we ([pkgforge](https://github.com/pkgforge)) also provide [appimages](https://github.com/pkgforge/pkgcache/tree/main) etc.
So, as long as you support the static binaries, there's no need to add any other format.

> will made, for example steam.AppImage lowercased, as steam.appimage, and this causes wget to have errors

I relied on the naming schema that developers themselves publish as, all Github Releases have `.AppImage`, but this was a mistake.
I am fixing our infra to use lowercase everywhere.


> you can instead inject https://pkg.pkgforge.dev/$ARCH/$arg.AppImage having that hardcoded at the end.

I will be changing it to always be lowercase for everything, so if you have already done this, you will need to revert/adjust.


> Since you are more familiar with AppBundle and other formats..

All other formats require special handling, which is probably out of scope for AM. 
For instance, this is what needs to be done for  `FlatImage` integration: https://flatimage.github.io/docs/cmd/desktop/
And AppBundles need a daemon: https://github.com/xplshn/pelf/?tab=readme-ov-file#drawbacks


Honestly, without metadata, I don't see how it will be possible.
You are parsing the readme, but I wouldn't recommend it.
And even if you go about using the json, the file at minimum will be ~ 5 MB, so for every request you would be parsing a 5-10 MB file.
Soar/dbin gets around this by storing the json locally to use it as their database, which is not possible for AM.

I don't recommend writing any code, as I am going to change some things.
I will comment here once I am done, on how to do it using grep/sed, if you really need it, but don't keep your hopes high.

**Comment by 00sapo** [comment]
> > that said, I'll allow only appimages
> 
> I believe @00sapo wanted the static binaries 

Exactly: most of the desktop apps are already available in AM, but pkgforge also provides many CLI tools that would extend the usability of AM

> And even if you go about using the json, the file at minimum will be ~ 5 MB, so for every request you would be parsing a 5-10 MB file. Soar/dbin gets around this by storing the json locally to use it as their database, which is not possible for AM.

It's not clear to me why AM can't store its own db. We already have a list of metadata in `.local/share/AM`.

`awk` may be better suited (even if slower) to parse jsons with standard unix tools, because awk is "history"-aware (i.e. it can match a particular line only if it comes after another). `grep` may be used for this as well, but it will probably require multiple calls that may make it slower.

I'm totally new to AM development, but if @ivan-hc tells me what we need to parse from the jsons, I look for some fast way of parsing them and storing in a file (e.g. `.local/share/AM/pkgforge`)

**Comment by Samueru-sama** [comment]
Sorry I got confused with the static bins and the toolpacks extras repo, the later is what I was discussing with Ivan.

**Comment by ivan-hc** [comment]
> I'm totally new to AM development, but if @ivan-hc tells me what we need to parse from the jsons, I look for some fast way of parsing them and storing in a file (e.g. `.local/share/AM/pkgforge`)

@00sapo It doesn't matter if it's in JSON or MD format, for me the important thing is that through common SHELL tools I can get the information I want to provide a message or a source that the user can refer to in case of problems (option `-a`, the one to show info on apps).

> > I believe @00sapo wanted the static binaries
> 
> Exactly: most of the desktop apps are already available in AM, but pkgforge also provides many CLI tools that would extend the usability of AM
> 
> > And even if you go about using the json, the file at minimum will be ~ 5 MB, so for every request you would be parsing a 5-10 MB file. Soar/dbin gets around this by storing the json locally to use it as their database, which is not possible for AM.
> 
> It's not clear to me why AM can't store its own db. We already have a list of metadata in `.local/share/AM`.

Same. It is not clear to me why it cannot have installation scripts and items to list in a compatible way as are the standards I have established, see option `newrepo`, here https://github.com/ivan-hc/AM/blob/main/docs/guides-and-tutorials/newrepo.md

Since many elements are already available in the main database, it would be enough to add the missing ones, even just through an external database but with compatible scripts (see option `-t`, here https://github.com/ivan-hc/AM/blob/main/docs/guides-and-tutorials/template.md ).

Automating the creation of such scripts would be very easy if you set the right parameters. See for example how this function creates scripts on the fly for Toolpack AppImages (this is what I worked for this issue, now it is in the "dev" branch of this repo):
```
_toolpack_handler() {
	mkdir -p "$CACHEDIR/toolpack"
	curl -Ls "$AMREPO"/templates/AM-SAMPLE-AppImage > "$CACHEDIR/toolpack/$arg.toolpack" || exit 1
	sed -i "s#APP=SAMPLE#APP=$arg#g; \
	s#FUNCTION#curl -Ls https://pkg.pkgforge.dev/$ARCH/$arg.AppImage.version#g; \
	s#wget \"\$version\"#wget https://pkg.pkgforge.dev/$ARCH/$arg.AppImage#g" "$CACHEDIR/toolpack/$arg.toolpack"
	chmod a+x "$CACHEDIR/toolpack/$arg.toolpack"
	arg="$CACHEDIR/toolpack/$arg.toolpack"
}
```
And see how the `-e` option creates a script on the fly for random AppImage packages on github (this is officially available in AM, from months):
```
		mkdir -p "$CACHEDIR/extra"
		curl -Ls "$AMREPO"/templates/AM-SAMPLE-AppImage > "$CACHEDIR/extra/$EXTERNAL_APP_NAME" || exit 1
		sed -i "s#APP=SAMPLE#APP=$EXTERNAL_APP_NAME#g" "$CACHEDIR/extra/$EXTERNAL_APP_NAME"
		API_GITHUB_REPO="https://api.github.com/repos/$USER_PROJECT/releases"
		sed -i "s#REPLACETHIS#$USER_PROJECT#g" "$CACHEDIR/extra/$EXTERNAL_APP_NAME"
		q="'"
		if [ "$ARCH" = "x86_64" ]; then
			FILTER=' | grep -vi "i386\\|i686\\|i586\\|i486\\|aarch64\\|arm64\\|armv7l"'
		elif [ "$ARCH" = "i686" ]; then
			FILTER=' | grep -ie "i386\\|i686\\|i586\\|i486" '
		elif [ "$ARCH" = "aarch64" ]; then
			FILTER=' | grep -ie "aarch64\\|arm64" '
		fi
		FUNCTION='curl -Ls '"$API_GITHUB_REPO"' | sed '"$q"'s/[()",{} ]/\\n/g'"$q"' | grep -oi "https.*mage$"'"$FILTER"' | head -1'
		sed -i "s#FUNCTION)#$FUNCTION)#g" "$CACHEDIR/extra/$EXTERNAL_APP_NAME"
```

and the Toolpack project already has all the elements needed to interact with the `-t` option via automated processes. In the end, creating a script for a static binary is not difficult, just follow the instructions https://github.com/ivan-hc/AM/blob/main/docs/guides-and-tutorials/template.md#option-two-archives-and-other-programs

**Comment by Azathothas** [comment]
>  I will comment here once I am done, on how to do it using grep/sed, if you really need it, but don't keep your hopes high.

It works, but it's just as fragile as parsing the Markdown files. And it still doesn't solve the problem of downloading heavy files first, just to get links/information to download another file.

> It's not clear to me why AM can't store its own db. We already have a list of metadata in .local/share/AM.

I asked @Samueru-sama, and I was told that, that Dir is used to store the list of apps for AM itself, it is not used for storing third party metadata. 

> awk may be better suited (even if slower) to parse jsons with standard unix tools

Perhaps, then parsing `TOML` or `YAML` would be easier.

> Same. It is not clear to me why it cannot have installation scripts and items to list in a compatible way as are the standards I have established, see option newrepo

For that to work, the `newrepo` needs to be in the same format as `AM`. This, unfortunately, isn't possible for us as we already have our own standards & specs.

> how this function creates scripts on the fly for Toolpack AppImages

I have changed things, so everything is now lowercased. `tr` could be used to convert all user input to lowercase, or the value can also be hard-coded as it's already been done: `$arg.AppImage` --> `$arg.appimage`

> In the end, creating a script for a static binary is not difficult, just follow the instructions

The actual problem is, how will `AM` get the URLs to download them in the first place.
`AM`, sure, could hard-code `https://bin.pkgforge.dev/${ARCH}/$BIN` but it won't have access to the [`Baseutils`](https://bin.ajam.dev/x86_64_Linux/Baseutils/) Dir & neither will it be able to get any other information.


---

I see these solutions:
1. Start including `JQ` as hard dependency for people who want to use `AM` for `pkgforge/toolpacks`. `AM` can just include this information in the README or have a `command -v jq` check for people who use that option and tell them to install `jq`. (This will still download 5-10 MB of JSON file) 
2. Modify the `.local/share/AM` to support third party metadata files, and use this dir to cache metadata from `pkgforge` or any other similar projects that need a `db` to work. This will probably require significant code changes.
@xplshn had made a full pkg manager just using `posix-tools` some time ago: https://github.com/xplshn/Handyscripts/blob/master/bdl (This contains outdated metadata parsers), so it is certainly possible.
3. Only add the `appimages` from `pkgforge`, this was already done, but it didn't resolve @00sapo's original issue. There already exists [`Soar`](https://github.com/pkgforge/soar) which @00sapo can use to get access to all of `pkgforge`'s [binary](https://github.com/Azathothas/Toolpacks)/[package](https://github.com/pkgforge/pkgcache) cache.

If there are more requests from more people, this could be revisited and reconsidered if it's worth it to invest significant time to add support for `pkgforge`.

**Comment by ivan-hc** [comment]
@Azathothas if you keep this file as is I can definitelly solve this issue and also the `-a` thing

https://raw.githubusercontent.com/Azathothas/Toolpacks/refs/heads/main/x86_64_Linux/README.md

**Comment by xplshn** [comment]
`dbin` is already in the repos. Why not just use `dbin findurl $PROGRAM` to get the URL of the program and use that to download it?
You can also use DBIN_INSTALL_DIR and just use `dbin add/install`.
`dbin findurl` is a hidden command that `dbin` has, it simply returns the actual URL of binary requested after having checked the meatadata (it also has fallbacks)

**Comment by xplshn** [comment]
> since the appImage is solved, now we need to finalize for all the other programs... and honestly, I have no idea how to start. Since you are more familiar with AppBundle and other formats... you could start a study on the changes I made in the "dev" branch and start adding the missing code.
> 
> https://github.com/ivan-hc/AM/pull/1082/files

I'm the author of the AppBundle format and tooling, integration is fairly easy to do. You can use the following flags if you already handle integration for AppImages:
`--pbundle_link`, this command is provided by the runtime and lets you run any command but from within the AppBundle. Its very powerful. Let's you do this, for example:
`--pbundle_link sh -c "cp \$SELF_TEMPDIR/.DirIcon ~/.local/share/icons/firefox.png"`
You also have the following commands:
`--pbundle_desktop | base64 -d`
`--pbundle_pngIcon | base64 -d`
`--pbundle_appstream | base64 -d`
`--pbundle_list`

You can check the exact build-date of an AppBundle via `--pbundle_help`.

I'm working on AppBundleHUB, which should provide metadata for the AppBundles it generates.
If you've looked at any of the recipes, you'll realice they're mostly one-liners like the following:
`pelfCreator -m "xplshn" -n "org.chromium.Chromium" -p "chromium" -e "chromium.desktop"`, when they include a `-e` that ends in `.desktop`, that means that the produced AppBundle will have an icon & desktop file. (if you don't like this automatic method of building/doing things, you can use the pelf-dwfs tool to create AppBundles of any AppDir, or the bs2appbundle tool to convert conty-based AppImages to AppBundle and it also supports normal AppImages.)

Anyhow. I'm open to suggestions :)

**Comment by ivan-hc** [comment]
Thanks @xplshn , I'm interested in alternative formats too

For now its better I take care of this issue first, then I'll work on the support for AppBundles... maybe in another issue.

**Comment by xplshn** [comment]
> Thanks @xplshn , I'm interested in alternative formats too
> 
> For now its better I take care of this issue first, then I'll work on the support for AppBundles... maybe in another issue.

I'm willing to make my dbin-metadata repo produce hourly text files in a format that you find easily parsable. dbin-metadata just contains `pkgforge`'s metadata, but I made the `pkg` key represent the actual place of the file in the repo. That's basically it. I can make another little converter/parser in Go just for AM to be able to retrieve easily parsable metadata from there.

**Comment by ivan-hc** [comment]
Progress on this issue resolution is suspended "until further notice". I am waiting for a "reliable" Markdown file to be provided as promised by @Azathothas in pvt

Anyway, my workflow will be similar to what already happens in `am -l --appimages`, I just need a markdown file like the one I linked here https://github.com/ivan-hc/AM/issues/1079#issuecomment-2462833802 and from there I'll have all stuff I need to download, install, get info, version, list... and everything.

I was already starting (today I released two version of "AM" in 2 hours... why not a third one?) but for now it's better that I stop.

@00sapo scusa se ancora non ho risolto, dovevo fare mente locale sul dove e come procedere per lavorare proprio su quella lista, e ora ho le idee un pò più chiare.

I will pin my "notes" about the function I'm working on in the next comment... just as a reminder for me.

**Comment by xplshn** [comment]
> Progress on this issue resolution is suspended "until further notice". I am waiting for a "reliable" Markdown file to be provided as promised by @Azathothas in pvt
> 
> Anyway, my workflow will be similar to what already happens in `am -l --appimages`, I just need a markdown file like the one I linked here [#1079 (comment)](https://github.com/ivan-hc/AM/issues/1079#issuecomment-2462833802) and from there I'll have all stuff I need to download, install, get info, version, list... and everything.
> 
> I was already starting (today I released two version of "AM" in 2 hours... why not a third one?) but for now it's better that I stop.
> 
> @00sapo scusa se ancora non ho risolto, dovevo fare mente locale sul dove e come procedere per lavorare proprio su quella lista, e ora ho le idee un pò più chiare.
> 
> I will pin my "notes" about the function I'm working on in the next comment... just as a reminder for me.

Not even if I'm the one to provide the markdown file? Also, why Markdown? Could be any format (even custom) that you request. (i'm bored xd)

**Comment by ivan-hc** [comment]
> Not even if I'm the one to provide the markdown file? Also, why Markdown? Could be any format (even custom) that you request. (i'm bored xd)

I already talked with @Azathothas on Discord, in brief, this is my workflow. The option `am -l --appimages` works like this to show only the AppImages. It is something I already know how to handle. That's all.

**Comment by xplshn** [comment]
> > Not even if I'm the one to provide the markdown file? Also, why Markdown? Could be any format (even custom) that you request. (i'm bored xd)
> 
> I already talked with @Azathothas on Discord, in brief, this is my workflow. The option `am -l --appimages` works like this to show only the AppImages. It is something I already know how to handle. That's all.

Oh, okay. Good to know then. Parsing the JSON and converting it to markdown is also possible.

**Comment by ivan-hc** [comment]
What a funny thing... I have tested my function on `7z`, and this is what happened

![Istantanea_2024-11-07_19-55-07 png](https://github.com/user-attachments/assets/4ec81f11-2bb1-40be-9de1-374455b55d5b)

**Comment by ivan-hc** [comment]
OK, the following function will be added to the install.am module
```
# This function is needed to handle Toolpacks
_toolpack_handler() {
	mkdir -p "$CACHEDIR/toolpack"
	toolpack_readme="https://raw.githubusercontent.com/Azathothas/Toolpacks/refs/heads/main/${ARCH}_Linux/README.md"
	toolpack_version="curl -Ls "$toolpack_readme" | grep -i \"^| $arg |\" | tr '\"><[]()| ' '\n' | grep . | head -9 | tail -1"
	curl -Ls "$AMREPO"/templates/AM-SAMPLE-Archive > "$CACHEDIR/toolpack/$arg.toolpack" || exit 1
	sed -i "s#APP=SAMPLE#APP=$arg#g; s#FUNCTION#$toolpack_version#g; \
	s#wget \"\$version\"#wget https://bin.ajam.dev/${ARCH}_Linux/$arg#g" "$CACHEDIR/toolpack/$arg.toolpack"
	sed -e '/\[ -e/ s/^#*/#/' -i "$CACHEDIR/toolpack/$arg.toolpack" 2>/dev/null
	chmod a+x "$CACHEDIR/toolpack/$arg.toolpack"
	arg="$CACHEDIR/toolpack/$arg.toolpack"
}
```
and it will be handled by this step
```
				if echo "$arg" | grep -q "/"; then
					if test -f "$arg" 2> /dev/null; then
						_install_local_script
					else
						echo " 💀 ERROR: the file \"$arg\" does NOT exist"
					fi
				elif echo "$@" | grep -q -- "--toolpack"; then
					if curl --output /dev/null --silent --head --fail "https://bin.ajam.dev/${ARCH}_Linux/$arg"  1>/dev/null; then
						_toolpack_handler
						_install_local_script
					else
						echo " 💀 ERROR: \"$arg\" does NOT exist in Toolpacks"
					fi
				elif curl --output /dev/null --silent --head --fail "$APPSDB"/"$arg"  1>/dev/null; then
					_install_normally
				else
					echo " 💀 ERROR: \"$arg\" does NOT exist in the database, see \"$AMCLI -l\""
				fi
```
this is what happens if I install `7z`

![Istantanea_2024-11-07_20-10-37 png](https://github.com/user-attachments/assets/65c700fe-586a-4642-b4ec-08fcadb645e2)

-----------------------------------------

I stop here... just the time @Azathothas have provided a good markdown.

When it will be ready, it will be a total game to handle like this
```
	toolpack_readme="https://raw.githubusercontent.com/Azathothas/Toolpacks/refs/heads/main/${ARCH}_Linux/README.md"
	toolpack_version="curl -Ls "$toolpack_readme" | grep -i \"^| $arg |\" | tr '\"><[]()| ' '\n' | grep . | head -9 | tail -1"
```
same for all info provided.

**Comment by ivan-hc** [comment]
Anyone who wants to test installations via AM can download the "install.am" module (remove the .txt extension) from here:

[install.am.txt](https://github.com/user-attachments/files/17668599/install.am.txt)

it's just a prototype... and a preview of what's to come.

**Comment by ivan-hc** [comment]
> Anyone who wants to test installations via AM can download the "install.am" module (remove the .txt extension) from here:
> 
> [install.am.txt](https://github.com/user-attachments/files/17668599/install.am.txt)
> 
> it's just a prototype... and a preview of what's to come.

@xplshn I'm not uploading it to "dev" yet because I want to wait for the new markdown from @Azathothas

**Comment by xplshn** [comment]
> > Anyone who wants to test installations via AM can download the "install.am" module (remove the .txt extension) from here:
> > [install.am.txt](https://github.com/user-attachments/files/17668599/install.am.txt)
> > it's just a prototype... and a preview of what's to come.
> 
> @xplshn I'm not uploading it to "dev" yet because I want to wait for the new markdown from @Azathothas

I said that I could provide you with one, generated from the JSON metadata, and I can assure you that it won't change. Since I always focus on compatibility with older `dbin` versions too. I've retained compat with 0.3, 0.4, 0.5, and we're now in the 0.6 version, even after all of the changes that Ajam's made to the metadata. (ofc, this is only possible since I pass all of the metadata through a converter, always, even if there's nothing to do but minify the metadata, this ensures that if something changes upstream, I can prepare and keep `dbin` working without any downtime.)

**Comment by ivan-hc** [comment]
@xplshn just talked with @Azathothas in pvt that have given a new list. this is how the function changes
```
# This function is needed to handle Toolpacks
_toolpack_handler() {
	mkdir -p "$CACHEDIR/toolpack"
	toolpack_readme="https://bin.ajam.dev/${ARCH}_Linux/AM.txt"
	toolpack_version="curl -Ls "$toolpack_readme" | grep -i \" $arg \" | tr '|' '\n' | cut -c 2- | grep . | tail -1"
	curl -Ls "$AMREPO"/templates/AM-SAMPLE-Archive > "$CACHEDIR/toolpack/$arg.toolpack" || exit 1
	sed -i "s#APP=SAMPLE#APP=$arg#g; s#FUNCTION#$toolpack_version#g; \
	s#wget \"\$version\"#wget https://bin.ajam.dev/${ARCH}_Linux/$arg#g" "$CACHEDIR/toolpack/$arg.toolpack"
	sed -e '/\[ -e/ s/^#*/#/' -i "$CACHEDIR/toolpack/$arg.toolpack" 2>/dev/null
	chmod a+x "$CACHEDIR/toolpack/$arg.toolpack"
	arg="$CACHEDIR/toolpack/$arg.toolpack"
}
```

**Comment by Azathothas** [comment]
Hi, a minor nitpick, please use the `pkgforge.dev` domain instead of my personal `ajam.dev` as I am migrating everything on the new domain.

**Comment by ivan-hc** [comment]
Strange things are happening. I noticed this while I'm working on `-a`.

If an app is missing from the database, the check will be done for toolpacks. This is the function that SHOULD detect if an argument exists
```
	'-a'|'about')
		[ -z "$2" ] && echo " USAGE: $AMCLI $1 [ARGUMENT]" && exit 1
		_online_check
		printf "%s\n" "-----------------------------------------------------------------------"
		[ ! -f "$AMCACHEDIR"/version-args ] && _check_version
		entries="$(echo "$@" | cut -f2- -d ' ')" # Removes first argument
		for arg in $entries; do
			if curl -o /dev/null -sIf "$AMCATALOGUEMARKDOWNS/${arg}.md" 1>/dev/null; then
				_about_page "$arg"
			elif curl -o /dev/null -sIf "$toolpack_sources/$ARCH/$arg"; then
				arg="$arg.toolpack"
				_about_page "$arg"
			elif echo "$arg" | grep -q ".toolpack$"; then
				_about_page "$arg"
			else
				printf " ERROR: %b not found\033[0m\n" "${RED}\"$arg\""
			fi
			printf "%s\n" "-----------------------------------------------------------------------"
		done
		;;
```
where the part of the function that handles this is
```
	elif echo "$package_name" | grep -q ".toolpack"; then
		toolpack_appname=$(echo "$arg" | sed 's/\.toolpack//g')
		if curl -o /dev/null -sIf "$toolpack_sources/${ARCH}/$toolpack_appname"; then
			toolpack_md=$(curl -Ls "$toolpack_readme" | grep -i "^| $toolpack_appname " | tr '|' '\n' | cut -c 2- | grep .)
			toolpack_about_description=$(echo "$toolpack_md" | awk -F: "NR==$toolpack_description")
			toolpack_about_site=$(echo "$toolpack_md" | awk -F: "NR==$toolpack_site")
			printf "\n%b\n\nSOURCE: %b\n" "$toolpack_about_description" "$toolpack_about_site" | fold -sw 72 | sed 's/^ //g; s/^/ /g'
		fi
	else
```
and the variables for toolpacks are these
```
# 3RD PARTY SOURCES --TOOLPACK
export toolpack_sources="https://bin.pkgforge.dev"
export toolpack_readme="$toolpack_sources/${ARCH}/AM.txt"
export toolpack_name="1"
export toolpack_description="2"
export toolpack_site="3"
export toolpack_dl="4"
export toolpack_ver="5"
```
well... this is what SHOULD happen if the argument is not detected by `curl -o /dev/null -sIf`

![Istantanea_2024-11-08_17-09-38](https://github.com/user-attachments/assets/e847d707-1b98-4ffd-9b96-72850505cfe6)

and this is what happens since I added Toolpack support

![Istantanea_2024-11-08_17-10-15](https://github.com/user-attachments/assets/cdf49636-4ed0-4e62-8e12-464bf338208d)

and this happens because `"$toolpack_sources/$ARCH/$arg"` redirects to another source, so `curl` have an answer somehow.

This must not happen!

After these tests I noticed that the same happens for the installation process. This is what it should do

![Istantanea_2024-11-08_17-34-37](https://github.com/user-attachments/assets/e10f21b3-b389-4216-af77-0fad5947d765)

...and this is what happens since I added Toolpack

![Istantanea_2024-11-08_17-35-23](https://github.com/user-attachments/assets/46d76cfb-1b76-4cf3-b4ec-ec4bf98971a2)

what the hell?!

**Comment by ivan-hc** [comment]
I have solved for both

**Comment by ivan-hc** [comment]
I think I have finished. I only need to add instructions in the "help" message.

@Azathothas @xplshn @00sapo @Samueru-sama you all can already test all this by running
```
am --devmode-enable
am -s
```
what is changed:
- option `-i` or `install`, use the flag `--toolpack` to install applications
- option `-q` or `query`, same, use the flag `--toolpack` to search something from the list
- option `-l` or `list`, same, use the flag `--toolpack` to create the list
- option `-a` or `about`, no flag at all... just search info of an app and if it is not in the AM database, the search will be done also among Toolpacks.

That's all. If you miss something, let me know.

I have done not many changes, as you can see https://github.com/ivan-hc/AM/pull/1096/files#diff-bcf89e62ca1f28f0c7b956b397c4d4fb7e6eefb32b0684b064000e0199dea554

**Comment by ivan-hc** [comment]
@Samueru-sama I made sure that the template to use for the on-the-fly script creation takes into account the application name: if it ends with `.appimage` the AppImage script will be used, otherwise the Archives script will always be used, which is fine for everything (even AppImages) but does not create .desktop files or icons, unless specified.

That said, if in the near future we want to support AppBundles and FlatImages or whatever... we need separate templates at https://github.com/ivan-hc/AM/tree/main/templates ... or at least a function that completes/patches the existing ones on the fly.

**Comment by ivan-hc** [comment]
@Azathothas in case you want to change something, please, stick to the variables I dedicated in APP-MANAGER
```
# 3RD PARTY SOURCES -- TOOLPACK
toolpack_repo="https://github.com/Azathothas/Toolpacks"
export toolpack_sources="https://bin.pkgforge.dev"
export toolpack_readme="$toolpack_sources/${ARCH}/AM.txt"
export toolpack_name="1"
export toolpack_description="2"
export toolpack_site="3"
export toolpack_dl="4"
export toolpack_ver="5"
```
- toolpack_repo is used in the "help" message, or somewhere else if needed
- toolpack_sources is used only in toolpack_readme, but it may be used somewhere else if needed
- toolpack_readme is used in the modules install.am and database.am for the options that require it
- all other variables are references to complete the following command
```
awk -F: "NR==$toolpack_variable_name"
```
to check from what part of the table it can have appname, version, description, download URL and original site.

If you leave the table with the same colums scheme, it is better... maybe you can change toolpack_sources and toolpack_readme if needed.

**Comment by ivan-hc** [comment]
I'll wait your feedback before I merge to "main".

If you have something to say, say it now... "or forever hold your peace" as they say at weddings.

**Comment by Azathothas** [comment]
Hi,
It looks good to me.
I won't be changing anything, even if I do, the way that AM.txt is generated will ensure backwards compatibility.
Maybe the original author of the issue @00sapo has anything to say...?

**Comment by ivan-hc** [comment]
To install a package it is enough to use the flag `--toolpack` or an extension `.toolpack`
```
am -i --toolpack {PROGRAM}
am -i {PROGRAM}.toolpack
```
same for local usage
```
am -i --toolpack --user {PROGRAM}
am -i --user {PROGRAM}.toolpack
appman -i --toolpack {PROGRAM}
appman -i {PROGRAM}.toolpack
```

**Comment by 00sapo** [comment]
Great! Thanks! It works for me!

I only have one additional feature I would like to have: I use `appman -f` to export the list of installed applications, so I can reinstall them in case I need. The problem now is tht if I use `appman -i --toolpack lazygit`, then `appman -f` will show lazygit among the installed apps and with no way of understanding if I need to use toolpack repo or not when reinstalling... is there any way to show `lazygit.toolpack` in the list printed by `-f`?

**Comment by ivan-hc** [comment]
> I only have one additional feature I would like to have: I use `appman -f` to export the list of installed applications, so I can reinstall them in case I need. The problem now is that if I use `appman -i --toolpack lazygit`, then `appman -f` will show lazygit among the installed apps and with no way of understanding if I need to use toolpack repo or not when reinstalling... is there any way to show `lazygit.toolpack` in the list printed by `-f`?

I was already thinking about it yesterday, it would need a bodo to identify the origin of the installed apps from one or the other database... only that when we refer to "installed files" we take into account the files possibly "to be removed" (options `-r` and `-R`) and with the extension, it would not work. I would have to modify an additional module in that case. The removal options refer to `-f` if a program is not installed.

Maybe I should add another column to the table.

**Comment by ivan-hc** [comment]
This is a series of tests I've done

| 1 | 2 | 3 | 4 | 5 |
| - | - | - | - | - |
| ![1](https://github.com/user-attachments/assets/afce8013-0870-4cde-8c02-a6435a12b90a) | ![2](https://github.com/user-attachments/assets/8fcfe14f-0171-4cef-be90-af1dcb3edeab) | ![3](https://github.com/user-attachments/assets/2ede8d2b-3541-46d8-9c38-3787c3e30724) | ![4](https://github.com/user-attachments/assets/fa1fd8a7-b587-4352-9d02-c548cd3ef6a2) | ![5](https://github.com/user-attachments/assets/bbcb0bf1-7619-4029-9c6e-fd4c6b1af59b)

I like more the number 5 (less invasive and at the same time detailed), but I'd like to know your opinion

@00sapo @Azathothas @xplshn @Samueru-sama

**Comment by ivan-hc** [comment]
Mybe the number five looks better here

![Istantanea_2024-11-09_20-22-29](https://github.com/user-attachments/assets/e2fbd884-0938-4859-9ed3-509724eb3ae9)

**Comment by Samueru-sama** [comment]
5 is ok

**Comment by ivan-hc** [comment]
maybe lowercased looks better

![Istantanea_2024-11-09_20-31-28](https://github.com/user-attachments/assets/41393724-9095-4057-a1d5-90f9df71113d)

**Comment by ivan-hc** [comment]
I did two tests with the `-e` option to install external apps, and at one point I emptied the hidden .am-installer directory... here's what came out

![Istantanea_2024-11-09_21-02-44](https://github.com/user-attachments/assets/ea43e4a6-6b55-49ca-b7e0-3ead0bea9bfe)

the one without a file in that directory, come from an unknown source, the other have a file containing `404: Not Found`, like the ones installed via toolpack, so it come not from a database

**Comment by ivan-hc** [comment]
I think it is ready. If everything is good, I'll release it as "AM" 9.1 tomorrow.

**Comment by Samueru-sama** [comment]
> I think it is ready. If everything is good, I'll release it as "AM" 9.1 tomorrow.

I did some tests and everything seems good. 

For some weird reason it can't find the `ncdu` binary but besides that, all good.

**Comment by xplshn** [comment]
Number 5 looked the best IMO too

**Comment by ivan-hc** [comment]
> For some weird reason it can't find the `ncdu` binary but besides that, all good.

is it a toolpack? I have tried with `am -q --toolpack` without success. Maybe the online list is updating.

**Comment by ivan-hc** [comment]
I noticed that the list of toolpacks was of 3000 items, but many of these were duplicates

in this commit https://github.com/ivan-hc/AM/commit/810d0e25417e41e50e5c6db97a73b3d86fe3e2e5 I fix this issue. Now listed toolpacks are 2474 (i.e. less programs than AM).

**Comment by Samueru-sama** [comment]
> > For some weird reason it can't find the `ncdu` binary but besides that, all good.
> 
> is it a toolpack? I have tried with `am -q --toolpack` without success. Maybe the online list is updating.

Yes it is a toolpack

**Comment by ivan-hc** [comment]
This other commit https://github.com/ivan-hc/AM/commit/ce49beb25f1a94ad07e76ccb64c2d8fe31b519c0 allows the use of completion terms for bash/zsh completion

**Comment by ivan-hc** [comment]
> > > For some weird reason it can't find the `ncdu` binary but besides that, all good.
> > 
> > 
> > is it a toolpack? I have tried with `am -q --toolpack` without success. Maybe the online list is updating.
> 
> Yes it is a toolpack

@Samueru-sama @Azathothas `ncdu` is not in this list https://bin.ajam.dev/x86_64_Linux/AM.txt

**Comment by ivan-hc** [comment]
This other commit https://github.com/ivan-hc/AM/commit/3f0c6e9c6c0549d5fbddd433e27ccfb94364e2d0 allows to use a `--all` flag in `-l` and `-q` to list and search all apps from all supported databases

https://github.com/user-attachments/assets/f9d2b5ce-8692-4d40-b6bb-ee7c0f6df467

**Comment by ivan-hc** [comment]
I thin that this commit is the more interesting https://github.com/ivan-hc/AM/pull/1096/commits/ce17a678700193ca5eabca5da3db0a7327d4b683 , with the `--toolpack` flag you can also install apps from AM if they are not present in Toolpacks

https://github.com/user-attachments/assets/c2febd0b-3570-4644-8b0e-82b7b5429b5d

**Comment by 00sapo** [comment]
Impressive! Thanks @ivan-hc !

![image](https://github.com/user-attachments/assets/94222b48-a32c-4601-9150-55d8e9ddcd10)

**Comment by 00sapo** [comment]
A better count:

```
~  appman -l --all | grep ◆ | wc -l                                                     
5079
```

**Comment by 00sapo** [comment]
And better:
```
❯ appman -l --all | awk '/◆/{print $2}' | sort | uniq | wc -l                                 
4965
```

**Comment by ivan-hc** [comment]
i conti non tornano

**Comment by ivan-hc** [comment]
this is the function that give the total number of programs available
```
_list_apps_all_msg() {
	_list_msg_head
	AVAILABLE_TOTAL_PROGRAMS=$(sort "$AMDATADIR/$ARCH-apps" "$AMCACHEDIR/$ARCH-toolpack" | uniq | grep -e "$" -c)
	echo " LIST OF THE $AVAILABLE_TOTAL_PROGRAMS PROGRAMS AVAILABLE IN ALL OUR SUPPORTED DATABASES:"
	_list_msg_footer
}
```
why it shows 5068 programs and your commit less?

**Comment by 00sapo** [comment]
> this is the function that give the total number of programs available
> 
> ```
> _list_apps_all_msg() {
> 	_list_msg_head
> 	AVAILABLE_TOTAL_PROGRAMS=$(sort "$AMDATADIR/$ARCH-apps" "$AMCACHEDIR/$ARCH-toolpack" | uniq | grep -e "$" -c)
> 	echo " LIST OF THE $AVAILABLE_TOTAL_PROGRAMS PROGRAMS AVAILABLE IN ALL OUR SUPPORTED DATABASES:"
> 	_list_msg_footer
> }
> ```
> 
> why it shows 5068 programs and your command less?

My command is imprecise because it counts the installed apps

**Comment by ivan-hc** [comment]
> My command is imprecise because it counts the installed apps

so it should show more lines, not 100 less

**Comment by ivan-hc** [comment]
you are right... I forget to jump counting "metapackages"

**Comment by 00sapo** [comment]
But when using `-l` I find more packages. I find this:

![image](https://github.com/user-attachments/assets/e77b4ec4-6fa7-47e2-9c2d-df77bf39d10e)

![image](https://github.com/user-attachments/assets/1ad37728-5581-4f34-bc78-70068069501f)

![image](https://github.com/user-attachments/assets/70d30bae-2ae6-4797-8050-fbb86fabc3c1)

![image](https://github.com/user-attachments/assets/414f477e-a338-4164-a9db-4bac09c9135a)

**Comment by ivan-hc** [comment]
https://github.com/ivan-hc/AM/pull/1096/commits/bb0638f57f8f7e50f3574c64d39483fe27cd7bf9

there are 79 items I don't include in total count, I call them "metapackages", i.e. many scripts with th exact same content to hel installing programs usinga keyword easier to remember, for example to install `platform-tools` you can use
```
am -i adb
```
or
```
am -i fastboot
```
because there are three scripts for `platform-tools` with the same content but three different names.

`kdegames` has more than 40 scripts, run `am -q kdegames` to see them. Same for `kdeutils` and `node` (the latter has one other script, `npm`).

---------------------------

Also, htere are some items that start with `ffwa-`, they are Firefox profiles, they are 9 in total. Also them are not included in the count.

-------------------------

so in total they are 79 "duplicated" I don't include in the total count.

Try now with `am -l --all`, the apps should be less than 5000 in total

**Comment by ivan-hc** [comment]
Now `am -l --all` lists 4989 programs, but `am -l --all | awk '/◆/{print $2}' | sort | uniq | wc -l` shows 4965

**Comment by 00sapo** [comment]
> Try now with `am -l --all`, the apps should be less than 5000 in total

It is 4989, indeed. But my command still shows different numbers (same as the screenshots above).

**Comment by 00sapo** [comment]
> Now `am -l --all` lists 4989 programs, but `am -l --all | awk '/◆/{print $2}' | sort | uniq | wc -l` shows 4965

exactly

**Comment by ivan-hc** [comment]
why? :thinking:

**Comment by xplshn** [comment]
@ivan-hc this is what you call "deduplication", the original metadata uses the binary names, and in the repo, there exists sub-folders for different implementations of the same tool, for example, on the repo, there's `busybox/whoami`, `coreutils/whoami`, and so on.

But on the `dbin`-flavored metadata, the `name` field accounts for this and in fact reflects the actual download_url.

It looks like this:
![image](https://github.com/user-attachments/assets/9af3d6a6-a76b-4680-aa23-98b5884c644f)
![image](https://github.com/user-attachments/assets/990f7635-c469-4b28-950e-0709ed6771f8)

**Comment by ivan-hc** [comment]
ok, thanks @xplshn

**Comment by ivan-hc** [comment]
I think many people want to see the total items when closing the list

![Istantanea_2024-11-10_02-56-44 png](https://github.com/user-attachments/assets/0b6064a5-54a2-4d7f-bf4e-a4ad00549614)

**Comment by Azathothas** [comment]
> ncdu is not in this list https://bin.ajam.dev/x86_64_Linux/AM.txt

It has been fixed in: https://pkg.pkgforge.dev/x86_64/AM.txt (https://github.com/pkgforge/pkgcache/commit/af5067dc11600634f6d9d5ecaf91fd5c2475f8c5)
```bash
❯ curl -qfsSL "https://pkg.pkgforge.dev/x86_64/AM.txt" | grep -i 'ncdu'
| ncdu | disk usage analyzer | https://dev.yorhel.nl | https://bin.pkgforge.dev/x86_64/ncdu | 5decfa473168 |
```

**Comment by ivan-hc** [comment]
Hey guys, get ready to run
```
am --devmode-disable
```
...to exit the developer mode. It's time to merge!

Get ready for AM 9.1 stable!

**Comment by ivan-hc** [comment]
https://github.com/ivan-hc/AM/releases/tag/9.1

**Comment by ivan-hc** [comment]
Folks, I'm sorry for this necrobumping... I have prepared a commit into a parallel branch https://github.com/ivan-hc/AM/compare/main...toolpacks-without-appimages

In brief, a lot of AppImages listed among the Toolpacks are duplicates of the existing ones in the AM repository, so this change prevents the installation and the listing of programs with name ending with `.appimage`... for the ones that are missing, we can dedicate a script for AM, so AM can manage the AppImages, while static binaries, appbundles and flatimages are provided by Toolpacks.

This is the list of Toolpacks without AppImages

![Istantanea_2024-11-10_22-44-10](https://github.com/user-attachments/assets/07cbf4a3-705f-4584-be17-7bcbf406a9e0)

...do you agree with that :+1:  ? Or should I leave all as is :-1: ?

**Comment by xplshn** [comment]
I don't agree because most AppImages in AM are glibc-only, while most appimages in Toolpacks are meant to work in any system, but I guess its okay, I don't use AM because it requires `bash` anyways, so this change doesn't really affect me

**Comment by ivan-hc** [comment]
> I don't agree because most AppImages in AM are glibc-only, while most appimages in Toolpacks are meant to work in any system, but I guess its okay, I don't use AM because it requires `bash` anyways, so this change doesn't really affect me

I noticed that there are some appimages I own and that are already in the repo of AM, same for others that come from their upstream developers... this is why I asked.

These are all appimages listed in Toolpacks if you run `am -q --toolpack "\.appimage"`
```
 SEARCH RESULTS FOR "\.APPIMAGE":

 ◆ 86box.appimage : Emulator of x86-based machines.
 ◆ acreom.appimage : An IDE based Knowledge Base using Markdown  (AppImage).
 ◆ actual.appimage : A local-first personal finance app.
 ◆ advanced-passgen.appimage : Advanced Password Generator.
 ◆ android-platform-tools.appimage : Android SDK Platform-Tools (adb, fastboot & more).
 ◆ anydesk.appimage : Desktop sharing application, providing remote support and online meetings.
 ◆ bottles.appimage : An easy-to-use wineprefix manager to Run Windows software and games on Linux.
 ◆ brave.appimage : Privacy-oriented browser for Desktop and Laptop computers.
 ◆ browservice.appimage : Browse the modern web on historical browsers.
 ◆ cemu.appimage : Wii U emulator.
 ◆ cpu-x.appimage : A FOSS tool to gather information on CPU, motherboard and more.
 ◆ cursor.appimage : The AI Code Editor.
 ◆ czkawka.appimage : Multi functional app to find duplicates, empty folders, similar images, broken files etc..
 ◆ deadbeef-nightly.appimage : A Modular (Extensible with Plugins) Audio Player that can play & convert almost all Audio Formats (Nightly).
 ◆ deadbeef-stable.appimage : A Modular (Extensible with Plugins) Audio Player that can play & convert almost all Audio Formats (Stable).
 ◆ desktopeditors.appimage : Office suite with text, spreadsheet and presentation editors.
 ◆ discord.appimage : All-in-one cross-platform voice and text chat for gamers.
 ◆ dolphin-emu.appimage : A GameCube / Wii emulator.
 ◆ duckstation.appimage : A fast Simulator/Emulator of the Sony PlayStation(TM) Console.
 ◆ dunst.appimage : Lightweight and customizable notification daemon.
 ◆ firefox-dev.appimage : A web browser built from Firefox source tree (Developer Edition).
 ◆ firefox-esr.appimage : A web browser built from Firefox source tree (ESR).
 ◆ firefox-nightly.appimage : A web browser built from Firefox source tree (Nightly).
 ◆ freetube.appimage : An Open Source YouTube app for privacy.
 ◆ gimp.appimage : The GNU Image Manipulation Program.
 ◆ heroic.appimage : A Native GOG, Epic, and Amazon Games Launcher.
 ◆ i3.appimage : A Tiling Window Manager.
 ◆ inkscape.appimage : FOSS Vector Graphics Editor.
 ◆ inkscape-dev.appimage : FOSS Vector Graphics Editor.
 ◆ kdenlive.appimage : FOSS video editor, based on MLT Framework and KDE Frameworks.
 ◆ krita.appimage : A free and open source painting application.
 ◆ lite-xl.appimage : A lightweight, simple, fast, feature-filled, and extremely extensible Text Editor.
 ◆ microsoft-edge-dev.appimage : The web browser from Microsoft (Dev).
 ◆ microsoft-edge-stable.appimage : The web browser from Microsoft (Stable).
 ◆ mpv.appimage : 🎥 Command line video player.
 ◆ nvtop.appimage : GPU & Accelerator process monitoring for AMD, Apple, Huawei, Intel, NVIDIA and Qualcomm.
 ◆ obsidian.appimage : A powerful knowledge base that works on top of a local folder of plain text Markdown files.
 ◆ obs-studio.appimage : FOSS cross-platform screencasting and streaming app (AppImage
 ◆ olive-editor.appimage : Free open-source non-linear video editor.
 ◆ opera.appimage : Faster, safer and smarter web browser.
 ◆ polybar.appimage : A fast and easy-to-use tool for creating status bars.
 ◆ prismlauncher.appimage : Free, open source launcher for Minecraft.
 ◆ protonup-qt.appimage : Install and manage Proton-GE and Luxtorpeda for Steam and Wine-GE for Lutris.
 ◆ puddletag.appimage : Powerful, simple, audio tag editor for GNU/Linux.
 ◆ qbittorrent-enhanced.appimage : A qBittorrent for with Enhanced Features.
 ◆ rofi.appimage : A window switcher, application launcher and dmenu replacement.
 ◆ ryujinx.appimage : A simple, experimental Nintendo Switch emulator.
 ◆ signal-desktop.appimage : A private messenger for Windows, macOS, and Linux.
 ◆ siyuan.appimage : A privacy-first personal knowledge management system that supports complete offline usage.
 ◆ spotify.appimage : Play music from the Spotify music service.
 ◆ steam.appimage : A video game digital distribution service and storefront from Valve.
 ◆ vlc.appimage : Cross-platform media player and streaming server.
 ◆ vscode.appimage : Open source source code editor developed by Microsoft.
 ◆ wezterm.appimage : A GPU-accelerated cross-platform Terminal Emulator and Multiplexer.
 ◆ wps-office.appimage : Office suite, formerly Kingsoft Office.
 ◆ xemu.appimage : Original Xbox emulator.
 ◆ zettlr.appimage : A markdown editor for writing academic texts and taking notes.
```
@xplshn I remember I have already many of them, and I suppose that they are the ones owned by their respective packagers... the reason is not to have duplicates.

What are the missing ones in AM?

**Comment by ivan-hc** [comment]
Just checked and the only one I have not in the database is the appimage of VScode, since I have the original portable archive `code` from upstream... also, the one you named `desktopeditors.appimage` is `onlyoffice` in my repo.

**Comment by Azathothas** [comment]
All of the appimages that are added to [`pkgcache`](https://github.com/pkgforge/pkgcache) are done so only after validating icons, desktops etc. 
Zsync is also embedded.
So these no longer resemble the source i.e official releases from "developers"
This is done so end users don't have to deal with broken appimages.

So if you want to only include official appimages, you should exclude the entire catalog. 

> What are the missing ones in AM?

None. And probably for a long time, since we will try adding appimages that AM already has first, before adding new ones.

> the one you named desktopeditors.appimage is onlyoffice in my repo.

We use repology & official app_id for naming schema. It's done this way so we can generate additional metadata easily.


> so AM can manage the AppImages, while static binaries, appbundles and flatimages are provided by Toolpacks.

Good Decision. I agree.
In the future, pkgforge's AppImage will not resemble anything from the official sources, so it can be misleading.

**Comment by ivan-hc** [comment]
> All of the appimages that are added to [`pkgcache`](https://github.com/pkgforge/pkgcache) are done so only after validating icons, desktops etc. Zsync is also embedded. So these no longer resemble the source i.e official releases from "developers" This is done so end users don't have to deal with broken appimages.

that's exactly the point. I built my database by relying on the work of the people who own the AppImage packages, whether they are working or not, and with the `-a` option and the download URLs inside the scripts... but especially for `-a`, which uses the catalog pages... users have direct contact with the original developer or packager. And they can ask him to add features or improve the quality of the packages. In this way, users push the developer to improve what he does.

If we decided to do everything ourselves, and without notifying the upstream developer, we would go against the very concept of "open-source community". Projects improve because there are users who report bugs, request features... and other users who contribute to improving the project. And that's exactly what happened for this AM release, 9.1. I wouldn't have gotten to release it if I hadn't received proposals.

So, your packages can be better, of course... I don't doubt it. But if I were the developer of these packages (and for a good part they ARE REALLY packages made by me), I would prefer you to make me a pull request. Also because in the metadata there is a reference to my repository as a source. I prefer to take my responsibilities, take my blame as well as the credit.

That said, I really thank you for the work you have done... but I will merge that commit into main. I have to do it, at this point.

**Comment by ivan-hc** [comment]
I've screwed up my projects so many times... but if they're better today it's because I've learned from them.

I guess I'm not the only one who thinks this way, right?

**Comment by Azathothas** [comment]
> whether they are working or not, and with the -a option and the download URLs inside the scripts... but especially for -a, which uses the catalog pages... users have direct contact with the original developer or packager. 

```bash
soar query steam.appimage #will display everything

soar log steam.appimage | less #will display the raw CI Build logs

soar inspect steam.appimage | less #will display the build script that was used

```
Even if the upstream doesn't provide logs or behaves shady, the way of extracting and checking the contents and repacking ensures that you will always have logs for any package from `pkgcache`


> But if I were the developer of these packages (and for a good part they ARE REALLY packages made by me), I would prefer you to make me a pull request. 

You can see my activity on GitHub, though you would have to filter by the date ranging at least a year ago
The tldr is, 99% of them don't care. The 1% who do, they are slow to respond, lack the technical skills or don't even want to fix it.
That leaves about 0.1% or less who actually behave like you are describing them, but this 0.1% means I waste 10–15 mins writing a detailed issue or PR for an app which will statistically will prove to be just a massive waste of time. 
The decision to fix stuff was born out of necessity not necessarily frustration, but frustration also makes up the majority of the rationale. 


> I've screwed up my projects so many times... but if they're better today it's because I've learned from my mistakes.
> I guess I'm not the only one who thinks this way, right?

I went from `Toolpacks` >> `Toolpacks Extras` >> `PkgForge`, and still can't migrate `Toolpacks` completely to `pkgforge` because I was dumb when I started it and screwed up so much that now it's impossible to migrate it.
We all make silly mistakes, it's part of the process.

**Comment by Samueru-sama** [comment]
> If we decided to do everything ourselves, and without notifying the upstream developer, we would go against the very concept of "open-source community". Projects improve because there are users who report bugs, request features... and other users who contribute to improving the project.

Ivan see this PR I made at helix fixing an issue with their appimage: https://github.com/helix-editor/helix/pull/11243

the lead dev didn't like that I used go-appimage because what if malware?! (**even though they were already using linuxdeploy**) so I went and modified the PR to even not use either of the two.

The PR is still open waiting...

I also have another with at [lite-xl](https://github.com/lite-xl/lite-xl/pull/1924) open and waiting, this is one is even worse because the current appimage no longer works.

**Comment by ivan-hc** [comment]
@Samueru-sama @Azathothas Guys, the more you come to tell me... all the more reason, YOU should take the place of the upstream packager/developer... with all the merits of the case.

I say this for you, you risk doing the work for those who do not deserve it.

And @Azathothas , precisely because you mentioned the slow responsiveness of certain developers... I am always here, actively, and at your disposal, as long as I am well... so why do the work twice? Help me to do it well instead, if you have better solutions.
